dotnet/src/Microsoft.AutoGen/AgentChat/Agents/AssistantAgent.cs

```diff
@@ -0,0 +1,340 @@
+    {
+        // TODO: feed the right Roles into the call
+        this.modelContext.AddRange(from message in item select message.ToCompletionClientMessage(ChatRole.User));
+
+        List<AgentMessage> innerMessages = [];
+
+        ChatOptions options = new()
+        {
+            ToolMode = ChatToolMode.Auto,
+
+            // TODO: We should probably just use the M.E.Ai.Abtraction types directly
+            // TODO: Should we cache this List? Also, why is this a list, rather than an Enumerable?
+            Tools = (from tool in this.Tools.Values select (AITool)tool.AIFunction).ToList()
+        };
+
+        ChatCompletion completion = await this.modelClient.CompleteAsync(this.modelContext, options, cancellationToken);
+
+        // TODO: Flatten traverses the message contents already; we are about to do it again: There's a better way.
+        this.modelContext.Add(completion.Message.Flatten());
+
+        // TODO: I am not sure this is doing what we want it to be doing: This checks that everything inside of the
+        // completion is a FunctionCall.
+
+        while (completion.Message.Contents.All(content => content is FunctionCallContent))
+        {
+            // TODO: A nicer API for this
+            List<FunctionCallContent> calls = completion.Message.Contents.Cast<FunctionCallContent>().ToList();
+
+            yield return new ChatStreamFrame
+            {
+                Type = ChatStreamFrame.FrameType.InternalMessage,
+                InternalMessage = ToAgentEvent(calls, this.Name)
+            };
+
+            List<Task<FunctionResultContent>> toolCallTasks = (from call in calls
+                                                               select this.toolManager.InvokeToolAsync(call, cancellationToken))
+                                                              .ToList();
+
+            // TODO: Enable streaming the results as they come in, rather than in a batch?
+            FunctionResultContent[] taskResult = await Task.WhenAll(toolCallTasks);
+            ToolCallExecutionEvent toolCallResult = ToAgentEvent(taskResult, this.Name);
+
+            this.modelContext.Add(toolCallResult.ToCompletionClientMessage());
+
+            yield return new ChatStreamFrame { Type = ChatStreamFrame.FrameType.InternalMessage, InternalMessage = toolCallResult };
+
+            List<Handoff> handoffs = (from FunctionResultContent result in taskResult
+                                      where this.toolManager.Handoffs.Contains(result.Name)
+                                      select (Handoff)this.toolManager.Tools[result.Name])
+                                     .ToList();
+
+            if (handoffs.Count > 1)
+            {
+                throw new InvalidOperationException($"Multiple handoffs detected: {String.Join(", ", from handoff in handoffs select handoff.Name)}");
```

Comment:
It's possible that mutiple handoffs can happen. In that case it's better to warn and choose the first one, and remind user to disable parallel tool calling.